android/src/main/kotlin/co/doneservices/callkeep/CallkeepSoundPlayerService.kt

```diff
@@ -0,0 +1,151 @@
+        vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+            val vibratorManager = this.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
+            vibratorManager.defaultVibrator
+        } else {
+            getSystemService(VIBRATOR_SERVICE) as Vibrator
+        }
+        audioManager = this.getSystemService(AUDIO_SERVICE) as AudioManager
+        when (audioManager?.ringerMode) {
+            AudioManager.RINGER_MODE_SILENT -> {
+            }
+            else -> {
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                    vibrator?.vibrate(VibrationEffect.createWaveform(longArrayOf(0L, 1000L, 1000L), 0))
+                } else {
+                    vibrator?.vibrate(longArrayOf(0L, 1000L, 1000L), 0)
+                }
+            }
+        }
+    }
+
+    private fun playSound(intent: Intent?) {
+        this.data = intent?.extras
+        val sound = this.data?.getString(
+                CallKeepBroadcastReceiver.EXTRA_CALLKEEP_RINGTONE_FILE_NAME,
+                ""
+        )
+        var uri = sound?.let { getRingtoneUri(it) }
+        if (uri == null) {
+            uri = RingtoneManager.getActualDefaultRingtoneUri(
+                    this@CallKeepSoundPlayerService,
+                    RingtoneManager.TYPE_RINGTONE
+            )
+        }
+        try {
+            mediaPlayer(uri!!)
+        } catch (e: Exception) {
+            try {
+                uri = getRingtoneUri("ringtone_default")
+                mediaPlayer(uri!!)
+            } catch (e2: Exception) {
+                e2.printStackTrace()
+            }
+        }
+    }
+
+    private fun mediaPlayer(uri: Uri) {
+        mediaPlayer = MediaPlayer()
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            val attribution = AudioAttributes.Builder()
+                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
+                    .setUsage(AudioAttributes.USAGE_NOTIFICATION_RINGTONE)
+                    .setLegacyStreamType(AudioManager.STREAM_RING)
+                    .build()
+            mediaPlayer?.setAudioAttributes(attribution)
+        } else {
+            mediaPlayer?.setAudioStreamType(AudioManager.STREAM_RING)
+        }
+        val assetFileDescriptor = applicationContext.getContentResolver().openAssetFileDescriptor(uri, "r")
+        if (assetFileDescriptor != null) {
+            mediaPlayer?.setDataSource(assetFileDescriptor)
+        } else {
+            mediaPlayer?.setDataSource(applicationContext, uri)
+        }
+        mediaPlayer?.prepare()
+        mediaPlayer?.isLooping = true
+        mediaPlayer?.start()
+    }
+
+    private fun getRingtoneUri(fileName: String) = try {
+        if (TextUtils.isEmpty(fileName)) {
+            RingtoneManager.getActualDefaultRingtoneUri(
+                    this@CallKeepSoundPlayerService,
+                    RingtoneManager.TYPE_RINGTONE
+            )
```

Comment:
This same code is written three times in this function.

Reply:
I'll make it DRYer